### DISCUSSION TITLE
Non-nullable Pointers/References

### DISCUSSION THREAD

**Discussion by nordlow** [MrSmith33/vox#2] [General]
Are non-nullable pointers and references planned for Vox?

**Comment by MrSmith33** [comment]
I cant say that they are 10/10 features like slices, so I would need to think about them first. If it is ref like in D, which is only available for `foreach` and function parameters, then it is 8/10. The only rough point is that ref is not really a type in D, which gives some headache.
Storing non-nullable reference in a struct is definitely tricky, because compiler needs to ensure that it is always initialized properly and it is convenient to the user.
What I can certainly say that I will 100% need something like `ref` for `foreach`.

**Comment by drathier** [comment]
Is it even possible to ensure a pointer is always initialized without some form of automatic gc/refcount or using refcaps/borrowing?

1. Alloc value on heap
2. Make a pointer to it
3. If (goldbach conjecture) free original value
4. Is the second pointer now invalid?

**Comment by MrSmith33** [comment]
yes, that might be impossible in general (or at least very tough)

**Comment by nordlow** [comment]
Incidentally...

https://github.com/dlang/dmd/blob/3132e116ef35acdb042bf0082df653b5fbf5e042/test/compilable/diag_access_new.d

Lots of code construct cases, such as loop-statements and switch-statements, remains but this is a proof of concept.

The analysis detects null deref of members of classes being either surely null (error) or maybe (warning or deprecation) null.

I'm planning on adding a `@notnull` qualifier for members that are initialized to non-null (typically via `new`) during construction of parenting class and that are never set to null before destruction of class. In the other words, the non-null pointer value of the `@notnull` member always outlives its parenting class.

The analysis uses a simple variant of forward data-flow analysis.

**Comment by nordlow** [comment]
> Is it even possible to ensure a pointer is always initialized without some form of automatic gc/refcount or using refcaps/borrowing?
> 
>     1. Alloc value on heap
> 
>     2. Make a pointer to it
> 
>     3. If (goldbach conjecture) free original value
> 
>     4. Is the second pointer now invalid?

The solution is to have a contract (such as an attribute `@notnull`) for when pointers and classes can be initialized.
